File: AlgoExpert/2. Medium/C++/Monotonic Array/Monotonic Array [O{(n)_(1)}].cpp
#include <iostream>

using namespace std;

// O(n) time | O(1) space - where n is the length of the array

bool isMonotonic(vector<int> array) {
	bool is_non_decreasing = true;
	bool is_non_increasing = true;
	for (int i = 1; i < array.size(); i++) {
		if (array[i] < array[i - 1]) {
			is_non_decreasing = false;
		}
		if (array[i] > array[i - 1]) {
			is_non_increasing = false;
		}
	}
	return is_non_decreasing || is_non_increasing;
}

File: AlgoExpert/2. Medium/Python/Monotonic Array/Monotonic Array [O{(n)_(1)}].py
# O(n) time | O(1) space - where n is the length of the array

def isMonotonic(array):
	if len(array) <= 2:
		return True
	
	direction = array[1] - array[0]
	for i in range(2, len(array)):
		if direction == 0:
			direction = array[i] - array[i - 1]
			continue
		if breaksDirection(direction, array[i - 1], array[i]):
			return False

	return True

def breaksDirection(direction, previousInt, currentInt):
	difference = currentInt - previousInt
	if direction > 0:
		return difference < 0
	return difference > 0